esp-wifi/src/compat/common.rs

```diff
@@ -34,102 +34,108 @@ static mut MUTEX_IDX_CURRENT: usize = 0;
+        Self {
+            dst,
+            capacity,
             len: 0,
         }
     }
 
-    pub unsafe fn from(c_str: *const u8) -> StrBuf {
-        let mut res = StrBuf {
-            buffer: [0u8; 512],
-            len: 0,
-        };
-
-        let mut idx: usize = 0;
-        while *(c_str.offset(idx as isize)) != 0 {
-            res.buffer[idx] = *(c_str.offset(idx as isize));
-            idx += 1;
-        }
+    pub fn len(&self) -> usize {
+        self.len
+    }
 
-        res.len = idx;
-        res
+    fn space(&self) -> usize {
+        self.capacity - self.len
     }
 
-    pub unsafe fn append_from(&mut self, c_str: *const u8) {
-        let mut src_idx: usize = 0;
-        let mut idx: usize = self.len;
-        while *(c_str.offset(src_idx as isize)) != 0 {
-            self.buffer[idx] = *(c_str.offset(src_idx as isize));
-            idx += 1;
-            src_idx += 1;
+    fn write(&mut self, byte: u8) {
+        unsafe {
+            self.dst.write(byte);
+            self.dst = self.dst.add(1);
         }
+    }
 
-        self.len = idx;
+    pub fn append_char(&mut self, c: char) {
+        let mut buf = [0u8; 4];
+        let char = c.encode_utf8(&mut buf);
+        self.append(char);
     }
 
     pub fn append(&mut self, s: &str) {
-        let mut idx: usize = self.len;
-        s.chars().for_each(|c| {
-            self.buffer[idx] = c as u8;
-            idx += 1;
-        });
-        self.len = idx;
-    }
+        // Write as many bytes as possible. We're writing a c string which means we don't have
+        // to deal with utf8 character boundaries, so this should be fine.
+        let len = s.len().min(self.space());
+        for byte in &s.as_bytes()[..len] {
+            self.write(*byte);
+        }
 
-    pub fn append_char(&mut self, c: char) {
-        let mut idx: usize = self.len;
-        self.buffer[idx] = c as u8;
-        idx += 1;
-        self.len = idx;
+        // vsnprintf's semantics: it counts unwritten bytes, too
+        self.len += s.len();
     }
 
-    pub unsafe fn as_str_ref(&self) -> &str {
-        core::str::from_utf8_unchecked(&self.buffer[..self.len])
+    pub fn append_byte(&mut self, b: u8) {
+        if self.space() >= 1 {
+            self.write(b);
+        }
+
+        // vsnprintf's semantics: it counts unwritten bytes, too
+        self.len += 1;
     }
 }
 
-impl Write for StrBuf {
+impl Write for StrWriter {
     fn write_str(&mut self, s: &str) -> Result<(), core::fmt::Error> {
         self.append(s);
         Ok(())
     }
 }
 
-pub unsafe extern "C" fn syslog(_priority: u32, format: *const u8, mut args: VaListImpl) {
-    #[cfg(all(feature = "wifi-logs", target_arch = "riscv32"))]
-    {
-        let mut buf = [0u8; 512];
-        vsnprintf(&mut buf as *mut u8, 511, format, args);
-        let res_str = StrBuf::from(&buf as *const u8);
-        info!("{}", res_str.as_str_ref());
+pub unsafe fn str_from_c<'a>(s: *const u8) -> &'a str {
```

Comment:
we could probably also use `core::ffi::CStr`

Reply:
Maybe, let's see :)